File: python_simulations/virtual_controller/test_joystick.py
#!/usr/bin/env python3
"""Test script for the virtual controller widgets"""

import sys
from PySide6.QtWidgets import QApplication, QMainWindow, QVBoxLayout, QWidget, QLabel
from PySide6.QtCore import Qt

from joystick_widget import JoystickWidget
from control_widget import ControlWidget


class TestWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("Virtual Controller Test")
        
        # Create central widget and layout
        central_widget = QWidget()
        self.setCentralWidget(central_widget)
        layout = QVBoxLayout(central_widget)
        
        # Create control widget
        self.control_widget = ControlWidget()
        layout.addWidget(self.control_widget)
        
        # Create labels to display values
        self.left_label = QLabel("Left: 0.00")
        self.left_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        layout.addWidget(self.left_label)
        
        self.right_label = QLabel("Right: 0.00")
        self.right_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        layout.addWidget(self.right_label)
        
        # Connect signals
        self.control_widget.valuesChanged.connect(self.on_values_changed)
        
        self.resize(900, 500)
    
    def on_values_changed(self, left, right):
        self.left_label.setText(f"Left (Forward): {left:.2f}")
        self.right_label.setText(f"Right (Turn): {right:.2f}")
        
        # Calculate differential drive velocities
        k1 = 1.0
        k2 = 0.5
        left_vel = k1 * left - k2 * right
        right_vel = k1 * left + k2 * right
        
        print(f"Left: {left:.2f}, Right: {right:.2f} -> "
              f"Wheel L: {left_vel:.2f}, Wheel R: {right_vel:.2f}")


def main():
    app = QApplication(sys.argv)
    window = TestWindow()
    window.show()
    sys.exit(app.exec())


if __name__ == '__main__':
    main()

File: qt_cpp/virtual_controller/test_joystick.cpp
/**
 * Test application for joystick and control widgets
 * 
 * Compile and run this to test the joystick widgets independently
 * from the main forward kinematics simulation.
 */

#include <QApplication>
#include <QMainWindow>
#include <QVBoxLayout>
#include <QLabel>
#include <QDebug>
#include "control_widget.h"

int main(int argc, char* argv[]) {
  QApplication app(argc, argv);

  QMainWindow window;
  window.setWindowTitle("Joystick Widget Test");

  QWidget* central = new QWidget();
  QVBoxLayout* layout = new QVBoxLayout(central);

  // Info label
  QLabel* info = new QLabel(
    "Left Joystick: Vertical, Range 0.0 to 1.0\n"
    "Right Joystick: Horizontal, Range -1.0 to +1.0"
  );
  info->setStyleSheet("padding: 10px; background-color: #f0f0f0;");
  layout->addWidget(info);

  // Control widget with two joysticks
  ControlWidget* control = new ControlWidget();
  layout->addWidget(control);

  // Status label to show current values
  QLabel* status = new QLabel("Left: 0.00 | Right: 0.00");
  status->setStyleSheet("padding: 10px; font-weight: bold;");
  layout->addWidget(status);

  // Connect to display values
  QObject::connect(control, &ControlWidget::valuesChanged,
    [status](double left, double right) {
      status->setText(QString("Left: %1 | Right: %2")
        .arg(left, 0, 'f', 2)
        .arg(right, 0, 'f', 2));
      qDebug() << "Values changed - Left:" << left << "Right:" << right;
    });

  window.setCentralWidget(central);
  window.resize(600, 400);
  window.show();

  return app.exec();
}

